src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/ClientCreds/JsonWebToken.cs

```diff
@@ -203,9 +210,36 @@ internal class JWTPayload
         [DataContract]
         internal sealed class JWTHeaderWithCertificate : JWTHeader
         {
+            private string _thumbPrint;
+            private string _certificate;
+
             public JWTHeaderWithCertificate(IClientAssertionCertificate credential)
                 : base(credential)
             {
+                _thumbPrint = this.Credential.Thumbprint;
+                X509CertificatePublicCertValue = null;
+
+                if (!_sendX5c)
+                    return;
+
+                //Check to see if credential is our implementation or developer provided.
+                if (credential.GetType().ToString() != "Microsoft.IdentityModel.Clients.ActiveDirectory.ClientAssertionCertificate")
+                {
+                    CallState.Default.Logger.Warning(null, "The implementation of IClientAssertionCertificate is developer provided and it should be replaced with library provided implmentation.");
+                    return;
+                }
+
+#if  NET45
+                if (credential is ClientAssertionCertificate cert)
+                {
+                    X509CertificatePublicCertValue = Convert.ToBase64String(cert.Certificate.GetRawCertData());
+                }
+#elif NETSTANDARD1_3
+                if (credential is ClientAssertionCertificate cert)
+                {
+                    X509CertificatePublicCertValue = Convert.ToBase64String(cert.Certificate.RawData);
+                }
+#endif
```

Comment:
can this difference be hidden in a helper/extension for each of the platforms?

Reply:
At the moment, I'm not sure. it may require a lot of changes.

Reply:
it will require a few changes.... an extension method in each of the libraries which extends the functionality of the Certificate type or you can do that as an extension to ether the type of the Certificate property  (or partial class) to the ClientAssertionCertificate. this is not a big deal, however I suggest creating a refactoring task for this one which we can include in dev, then we can work on it together